### PR TITLE
2D plots with non-regular bins

### LINKED ISSUE #972: Plotting multi-dimensional data with mask fails with exception
```python
import numpy as np
import scipp as sc
from scipp import Dim
from scipp.plot import plot

data = sc.DataArray(
    data=sc.Variable(dims=['pack', 'tube', 'straw', 'pixel'], values=np.random.rand(2,8,7,256)),
    coords={})

data += sc.Variable(dims=['pixel'], values=np.sin(np.linspace(0,3.14, num=256)))

data.masks['tube_ends'] = sc.Variable(dims=['pixel'], values=np.full(256, False))

plot(data)
```

### LINKED ISSUE #1112: 2D plotting with non-equal bin size does not work
```Py
N = 10
M = 5
x = np.arange(N+1)**2
y = np.arange(M+1)
z = np.random.rand(M, N).astype(np.float64)
d1 = sc.Dataset()
d1.coords['x'] = sc.Variable(['x'], values=x, unit=sc.units.m)
d1.coords['y'] = sc.Variable(['y'], values=y, unit=sc.units.m)
d1['Signal'] = sc.Variable(['y', 'x'], values=z, unit=sc.units.counts)
plot(d1)
```
should produce something like
Screenshot_2020-05-01_09-14-06
but instead it produces regular pixels
Screenshot_2020-05-01_09-22-15

### PR DESCRIPTION
**Pull request by nvaytet**
This makes 2D plots with non-evenly spaced bins possible (before they were just plotted as regular bins):
```Py
N = 10
M = 5
x = np.arange(N+1).astype(np.float)**2
y = np.arange(M+1).astype(np.float)
z = np.random.random([M, N])
d1 = sc.Dataset()
d1.coords['x'] = sc.Variable(['x'], values=x, unit=sc.units.m)
d1.coords['y'] = sc.Variable(['y'], values=y, unit=sc.units.m)
d1['Signal'] = sc.Variable(['y', 'x'], values=z, unit=sc.units.kg)
plot(d1)
```
Screenshot_2020-06-11_18-39-30

It works by resampling the data onto an image with square pixels that is used for display in the matplotlib `imshow`. The resolution of that image is based on the `dpi` set in the `config`.

The resampling is performed using `rebin`. As a result, when an original bin edge falls in the middle of an image pixel, this pixel will be assigned the mean of the two neighbouring bins. This becomes clear when you zoom in onto an image:
Screenshot_2020-06-11_17-44-50

Note that there is currently no callback on the canvas to resample again as we zoom in or out. So if you have a large 2D data slice (larger than ~800x533), you will be limited to 800x533 pixels. The image resolution is by default set by the `width` and `height` in the `config`. This can be adjusted using the `resolution` keyword argument, which can either be an integer (in which case the x and y size of the image are the same), or a dict: `{"x": 512, "y": 256}`.
We plan to add this dynamic resampling in the future. This will also alleviate the split-bin effect shown in the image above.

**Additional notes:**
- This PR removes plotting variances in 2d plots (this would probably only very rarely used anyway; see discussion below)
- The API for using `matplotlib` axes has been simplified slightly, since we no longer have axes for variances: 
  - Before: `plot(..., mpl_axes={"ax": myax0, "cax": myax1})`
  - After: `plot(..., ax=myax0, cax=myax1)`
- Also fixed an old issue with plotting ND data when masks are present (trying to slice a dim which is not present in the mask)

Fixes #1112 
Fixes #972

### CODE REVIEW

File: python/src/scipp/plot/plot_2d.py [hunk 1]
```diff
@@ -251,6 +258,56 @@ def update_axes(self):
...
+                sc.dtype.float64)
+
+        # Compute bin widths for normalization pre-rebin in order to retain the
+        # pixel values instead of spreading the data over the pixels.
+        self.xywidth["x"] = sc.Variable(
+            dims=self.xyrebin["x"].dims,
+            values=np.ediff1d(self.xyedges["x"].values) / self.image_dx)
+        self.xywidth["y"] = sc.Variable(
+            dims=self.xyrebin["y"].dims,
+            values=np.ediff1d(self.xyedges["y"].values) / self.image_dy)
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:308]: Here and above you added a lot of code, maybe it can go into a helper function? If you do that, you may even be able to get rid of the duplication, because you can just call the same function twice, once for x and once for y, with corresponding parameters.

File: python/src/scipp/plot/plot_2d.py [hunk 2]
```diff
@@ -282,20 +339,84 @@ def update_slice(self, change):
...
                 # At this point, after masks were combined, all their
                 # dimensions should be contained in the data_array.dims.
                 if self.params["masks"][self.name]["show"]:
                     mslice = mslice[val.dim, val.value]
             else:
-                button_dims[self.buttons[dim].value.lower() == "x"] = val.dim
+                # Get the dimensions of the dimension-coordinates, since
+                # buttons can contain non-dimension coordinates
+                button_dims[self.buttons[dim].value.lower() ==
+                            "x"] = self.slider_x[self.name][val.dim].dims[0]
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:351]: Which buttons is the referring to? Aren't buttons labels with dimensions, not coord names?

File: python/src/scipp/plot/plot_2d.py [hunk 3]
```diff
@@ -282,20 +339,84 @@ def update_slice(self, change):
...
+                elif dim == button_dims[1]:
+                    mslice_dims.append(self.xyedges["x"].dims[0])
+                else:
+                    mslice_dims.append(dim)
+            vslice.masks["all"] = sc.Variable(dims=mslice_dims,
+                                              values=mslice.values)
+
+        # The scaling by bin width and rebin operations below modify the
+        # variances in the data, so here we have to manually split the values
+        # and variances into separate Variables. We store them in a dict.
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:394]: Why does this imply that we have to split them? Do you need different error handling?

**Comment by SimonHeybrock** [comment]
IS this possibly (intentionally or unintentinally) also fixing #972?

File: python/src/scipp/plot/plot_2d.py [hunk 4]
```diff
@@ -251,6 +258,56 @@ def update_axes(self):
...
+        xdims = self.xyrebin["x"].dims
+        if not self.histograms[self.name][xdims[0]]:
+            self.xyedges["x"] = sc.Variable(
+                dims=xdims,
+                values=centers_to_edges(
+                    self.slider_x[self.name][xdims[0]].values),
+                unit=self.slider_x[self.name][xdims[0]].unit)
+        else:
+            self.xyedges["x"] = self.slider_x[self.name][xdims[0]].astype(
+                sc.dtype.float64)
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:289]: If you are asking if i can do this
```Py
N = 10
M = 5
vecs = []
for i in range(N):
    vecs.append(np.random.random(3))
d = sc.Dataset()
d.coords['x'] = sc.Variable(['x'],
                            values=vecs,
                            unit=sc.units.m,
                            dtype=sc.dtype.vector_3_float64)
d.coords['y'] = sc.Variable(['y'],
                            values=["a", "b", "c", "d", "e"],
                            unit=sc.units.m)
d["Signal"] = sc.Variable(['y', 'x'],
                          values=np.random.random([M, N]),
                          unit=sc.units.counts)
plot(d)
```
the answer is yes
Screenshot_2020-06-12_13-57-51

File: python/src/scipp/plot/plot_2d.py (hunk 2 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:351]: No the buttons are using the coordinate names: https://scipp.github.io/visualization/plotting-2d-data.html#Using-labels-along-some-axis

Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:351]: I mean the buttons used to flip the X-Y axes of the image

File: python/src/scipp/plot/plot_2d.py (hunk 3 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:394]: Yes i need different error handling, because when i do https://github.com/scipp/scipp/blob/non_regular_bins_2d_plot/python/src/scipp/plot/plot_2d.py#L413, the errors grow in quadrature

File: python/src/scipp/plot/plot_2d.py (hunk 2 above)
Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:351]: But that is something quite different: Labels are used for labeling the tics, but the dimension is the dimension, so buttons should maybe still be labeled with the dim?

File: python/src/scipp/plot/plot_2d.py (hunk 3 above)
Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:394]: But that is how it is supposed to be, no?

Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:394]: it is, but here i am trying to trick `rebin` into doing something it is not supposed to. i'm resampling an image just for the purposes of displaying it. if i don't do the split, the variances end up looking like this
Screenshot_2020-06-12_14-31-26
when they should actually be like this
Screenshot_2020-06-12_14-31-36

File: python/src/scipp/plot/plot_2d.py (hunk 2 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:351]: i think it would be counter-intuitive to ask for `axes=[y, somelabels]` and get a figure with `somelabels` as the x axis label, and then two buttons with `y` and `x` for their description. I think it would no be clear where the `x` came from.

File: python/src/scipp/plot/plot_2d.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:289]: it creates fake coordinates with integers, and this is what is being used behind the scenes

[the remark this answers is not in the document]
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:289]: that said, it does not behave well when you zoom in/out and you have vectors or string as tick labels, but this is also the case in `master`... i'll open an issue for this

File: python/src/scipp/plot/plot_2d.py (hunk 2 above)
Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:351]: But aren't we actually using the dimension coord for plotting the data, and the axes just for the labeling?

File: python/src/scipp/plot/plot_2d.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:289]: 👍

File: python/src/scipp/plot/plot_2d.py (hunk 3 above)
Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:394]: That may be, but how do we know which one is correct? We need to follow through the logic of scaling by bin width, and how that is scaling the variances. We can look it up in Mantid, it is implemented there correctly (for example somewhere in `HistogramData/`.

Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:394]: Mantid conversion from counts to frequencies (for variances): https://github.com/mantidproject/mantid/blob/7a904dff01c3b0dfef00ccea8d55f7b0715a077b/Framework/HistogramData/src/FrequencyVariances.cpp#L41.

Division by factor without variance used for division by bin width: https://github.com/scipp/scipp/blob/dde207e7d5ecab159c9dc27ed3d4fd6089cf27d2/core/include/scipp/core/value_and_variance.h#L140-L143.

I would conclude that the behavior should be correct if you just do the division by bin width, then rebin, then multiply by new width.

Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:407]: After discussion:

What happens the other way round, when you are down-sampling? This is not clear what the variance be in an image pixel whichi is made up from several array elements in the underlying data.

It seems that if we cannot come up with a rule that works the same for both upsampling and downsampling, there probably is a flaw somewhere in your design.

Also, we are not sure how useful it is to show the variances, and it complicates a lot of the plotting code.

The decision is thus to remove showing the variances for 2d plots. If a user wants to look at the variances, they will either show up as error bars in 1d plots, or they can extract the variances and place them as values in a new Variable

File: python/src/scipp/plot/plot_2d.py [hunk 5]
```diff
@@ -251,29 +226,81 @@ def update_axes(self):
...
+                unit=self.slider_x[self.name][ydims[0]].unit)
+        else:
+            self.xyedges["y"] = self.slider_x[self.name][ydims[0]].astype(
+                sc.dtype.float64)
+
+        # Compute bin widths for normalization pre-rebin in order to retain the
+        # pixel values instead of spreading the data over the pixels.
+        self.xywidth["x"] = sc.Variable(
+            dims=self.xyrebin["x"].dims,
+            values=np.ediff1d(self.xyedges["x"].values) / self.image_dx)
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:273]: What is the reason for using `ediff1` instead of simply `self.xyedges["x"][dim, 1:] - self.xyedges["x"][dim, :-1]`?

File: variable/rebin.cpp [hunk 6]
```diff
@@ -30,6 +30,8 @@ void rebin_non_inner(const Dim dim, const VariableConstView &oldT,
   // coord is 1D.
   int iold = 0;
   int inew = 0;
+  const bool is_bool = newT.dtype() == dtype<bool>;
+  // const bool is_bool = newT.dtype().index == std::type_index(typeid(bool));
```
Comment (SimonHeybrock) [inline variable/rebin.cpp:34]: commented code.

File: variable/test/rebin_test.cpp [hunk 7]
```diff
@@ -20,3 +20,57 @@ TEST(Variable, rebin) {
...
+};
+
+TEST_F(RebinMask2DTest, mask_weights_2d) {
+  const auto edges = makeVariable<double>(Dimensions{Dim::X, 5},
+                                          Values{1.0, 3.0, 4.0, 5.5, 6.0});
+  const auto expected = makeVariable<bool>(
+      Dimensions{{Dim::Y, 2}, {Dim::X, 4}},
+      Values{true, false, true, true, false, true, false, false});
+
+  const auto result = rebin(mask, Dim::X, x, edges);
```
Comment (SimonHeybrock) [inline variable/test/rebin_test.cpp:73]: Youm ade changes to `rebin_non_inner`, so may also test that case, i.e., rebin `Dim::Y`?

File: variable/rebin.cpp (hunk 6 above)
Comment (nvaytet) [reply variable/rebin.cpp:34]: yes this is still part of my debugging for the windows build...

Comment (SimonHeybrock) [reply variable/rebin.cpp:34]: Does it work if you put the namespace, i.e., `core::dtype<bool>`?

Comment (nvaytet) [reply variable/rebin.cpp]: No.
I’ve done this workaround, let me know if it is a bad idea https://github.com/scipp/scipp/pull/1198/commits/53894509fba8abeb71a4c718fc652094caa3ab43

File: core/include/scipp/core/dtype.h [hunk 8]
```diff
@@ -26,7 +26,8 @@ struct SCIPP_CORE_EXPORT DType {
   bool operator!=(const DType &t) const noexcept { return index != t.index; }
   bool operator<(const DType &t) const noexcept { return index < t.index; }
 };
-template <class T> inline DType dtype{std::type_index(typeid(T))};
+template <class T>
+SCIPP_CORE_EXPORT inline DType dtype{std::type_index(typeid(T))};
```
Comment (SimonHeybrock) [inline core/include/scipp/core/dtype.h:30]: Undo this change, since you are using a workaround anyway now?

File: python/src/scipp/plot/plot_2d.py [hunk 9]
```diff
@@ -91,95 +89,76 @@ def __init__(self,
...
         self.vminmax = {"vmin": vmin, "vmax": vmax}
         self.global_vmin = np.Inf
         self.global_vmax = np.NINF
+        if resolution is not None:
+            if isinstance(resolution, int):
+                self.image_resolution = {"x": resolution, "y": resolution}
+            else:
+                self.image_resolution = resolution
+        else:
+            self.image_resolution = 0.64 * config.plot.dpi / 96.0
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:105]: What are these magic numbers?

File: python/src/scipp/plot/plot_2d.py [hunk 10]
```diff
@@ -217,7 +196,6 @@ def update_axes(self):
         for dim, button in self.buttons.items():
             if self.slider[dim].disabled:
                 but_val = button.value.lower()
-                # xc = self.slider_x[self.name][dim].values
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:220]: commented code

File: python/src/scipp/plot/plot_2d.py [hunk 11]
```diff
@@ -91,95 +89,76 @@ def __init__(self,
...
+        self.members["colorbar"] = self.cbar
+        if self.params["masks"][self.name]["show"]:
+            self.im["masks"] = self.ax.imshow(
+                [[1.0, 1.0], [1.0, 1.0]],
+                extent=extent_array,
+                norm=self.params["values"][self.name]["norm"],
+                origin="lower",
+                interpolation="nearest",
+                aspect=self.aspect,
+                cmap=self.params["masks"][self.name]["cmap"])
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:157]: Near duplicate code (and parameters) to 20 lines above. -> helper function?

File: python/src/scipp/plot/plot_2d.py [hunk 12]
```diff
@@ -282,20 +294,63 @@ def update_slice(self, change):
...
+                    mslice_dims.append(self.xyedges["x"].dims[0])
+                else:
+                    mslice_dims.append(dim)
+            vslice.masks["all"] = sc.Variable(dims=mslice_dims,
+                                              values=mslice.values)
+
+        # Scale by bin width and then rebin in both directions
+        vslice *= self.xywidth["x"] * self.xywidth["y"]
+        vslice = sc.rebin(vslice, self.xyrebin["x"].dims[0], self.xyrebin["x"])
+        vslice = sc.rebin(vslice, self.xyrebin["y"].dims[0], self.xyrebin["y"])
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:350]: So are we *always* doing the rebin step, even if we have equal bins?

File: python/src/scipp/plot/plot_2d.py (hunk 10 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:220]: it's been removed, hasn't it?

Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:220]: Too little sleep. :D

File: python/src/scipp/plot/plot_2d.py (hunk 12 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:350]: yes, i will change to only rebinning when necessary

File: python/src/scipp/plot/plot_2d.py (hunk 9 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:105]: :+1: 
this is a remnant from experimentation. i will change back to using the config width and height.

**Comment by nvaytet** [comment]
Ah sorry i forgot to add notes to the list of breaking changes, will do that this morning
